In [32]:
from uplink import Consumer, get, post, returns, Query, QueryMap
from uplink import headers
from collections.abc import Mapping

| Component    | Example            | Uplink annotation    | Curl         |
| ------------ | ------------------ | -------------------- | ------------ |
| Path         | `/users/{user}`    | Path parameter       | URL segment  |
| Query params | `?sort=created`    | `Query` / `QueryMap` | `?key=value` |
| Body (JSON)  | `{"name": "repo"}` | `Body`               | `-d`         |


In [ ]:


@headers({                                       # @headers can be used as a class decorator for headers that need to be added to every request:
    "Accept": "application/vnd.github.v3.full+json",
    "User-Agent": "Uplink-Sample-App"
})                      
class Github(Consumer):

    @get('/repositories')
    def get_repos(self):
        """ List all public repos """

    # Path Parameters
    @returns.json    # @returns.json to directly return the JSON response, instead of a response object, we can use the key argument to strictly return the nested field.
    @get('/users/{user}')   
    def get_user(self, user:str):
        """ Get github user"""
        pass

    
    # Query Parameters
    @returns.json
    @headers({                        # You can set static headers for a method using the @headers decorator.
    "Accept": "application/vnd.github.v3.full+json",
    "User-Agent": "Uplink-Sample-App"
    })
    @get("/users/{user}/repos")   
    def get_user_repos(self, user: str, sortby: Query = "created"):
        """ List user repos """
        pass

    # QueryMap: useful for “catch-all” or complex query parameter combinations
    @returns.json
    @get('/search/repositories')
    def search_repos(self, query_params: QueryMap):
        """
        Search repos with multiple query params automatically mapped
        """
        pass

# By default, uplink uses Requests

client = Github(base_url="https://api.github.com/")

In [ ]:

client.get_repos().json()

[{'id': 1,
  'node_id': 'MDEwOlJlcG9zaXRvcnkx',
  'name': 'grit',
  'full_name': 'mojombo/grit',
  'private': False,
  'owner': {'login': 'mojombo',
   'id': 1,
   'node_id': 'MDQ6VXNlcjE=',
   'avatar_url': 'https://avatars.githubusercontent.com/u/1?v=4',
   'gravatar_id': '',
   'url': 'https://api.github.com/users/mojombo',
   'html_url': 'https://github.com/mojombo',
   'followers_url': 'https://api.github.com/users/mojombo/followers',
   'following_url': 'https://api.github.com/users/mojombo/following{/other_user}',
   'gists_url': 'https://api.github.com/users/mojombo/gists{/gist_id}',
   'starred_url': 'https://api.github.com/users/mojombo/starred{/owner}{/repo}',
   'subscriptions_url': 'https://api.github.com/users/mojombo/subscriptions',
   'organizations_url': 'https://api.github.com/users/mojombo/orgs',
   'repos_url': 'https://api.github.com/users/mojombo/repos',
   'events_url': 'https://api.github.com/users/mojombo/events{/privacy}',
   'received_events_url': 'https://ap

In [15]:
client.get_user(user="hackaholic")

{'login': 'hackaholic',
 'id': 3721438,
 'node_id': 'MDQ6VXNlcjM3MjE0Mzg=',
 'avatar_url': 'https://avatars.githubusercontent.com/u/3721438?v=4',
 'gravatar_id': '',
 'url': 'https://api.github.com/users/hackaholic',
 'html_url': 'https://github.com/hackaholic',
 'followers_url': 'https://api.github.com/users/hackaholic/followers',
 'following_url': 'https://api.github.com/users/hackaholic/following{/other_user}',
 'gists_url': 'https://api.github.com/users/hackaholic/gists{/gist_id}',
 'starred_url': 'https://api.github.com/users/hackaholic/starred{/owner}{/repo}',
 'subscriptions_url': 'https://api.github.com/users/hackaholic/subscriptions',
 'organizations_url': 'https://api.github.com/users/hackaholic/orgs',
 'repos_url': 'https://api.github.com/users/hackaholic/repos',
 'events_url': 'https://api.github.com/users/hackaholic/events{/privacy}',
 'received_events_url': 'https://api.github.com/users/hackaholic/received_events',
 'type': 'User',
 'user_view_type': 'public',
 'site_admi

In [17]:
client.get_user_repos(user="hackaholic", sortby="updated")

[{'id': 957396259,
  'node_id': 'R_kgDOORC1Iw',
  'name': 'AIML',
  'full_name': 'hackaholic/AIML',
  'private': False,
  'owner': {'login': 'hackaholic',
   'id': 3721438,
   'node_id': 'MDQ6VXNlcjM3MjE0Mzg=',
   'avatar_url': 'https://avatars.githubusercontent.com/u/3721438?v=4',
   'gravatar_id': '',
   'url': 'https://api.github.com/users/hackaholic',
   'html_url': 'https://github.com/hackaholic',
   'followers_url': 'https://api.github.com/users/hackaholic/followers',
   'following_url': 'https://api.github.com/users/hackaholic/following{/other_user}',
   'gists_url': 'https://api.github.com/users/hackaholic/gists{/gist_id}',
   'starred_url': 'https://api.github.com/users/hackaholic/starred{/owner}{/repo}',
   'subscriptions_url': 'https://api.github.com/users/hackaholic/subscriptions',
   'organizations_url': 'https://api.github.com/users/hackaholic/orgs',
   'repos_url': 'https://api.github.com/users/hackaholic/repos',
   'events_url': 'https://api.github.com/users/hackaholic/

In [22]:
client.search_repos(query_params={
    "q": "machine learning",
    "sort": "stars",
    "order": "desc",
    "per_page": 10
})

{'total_count': 951306,
 'incomplete_results': False,
 'items': [{'id': 45717250,
   'node_id': 'MDEwOlJlcG9zaXRvcnk0NTcxNzI1MA==',
   'name': 'tensorflow',
   'full_name': 'tensorflow/tensorflow',
   'private': False,
   'owner': {'login': 'tensorflow',
    'id': 15658638,
    'node_id': 'MDEyOk9yZ2FuaXphdGlvbjE1NjU4NjM4',
    'avatar_url': 'https://avatars.githubusercontent.com/u/15658638?v=4',
    'gravatar_id': '',
    'url': 'https://api.github.com/users/tensorflow',
    'html_url': 'https://github.com/tensorflow',
    'followers_url': 'https://api.github.com/users/tensorflow/followers',
    'following_url': 'https://api.github.com/users/tensorflow/following{/other_user}',
    'gists_url': 'https://api.github.com/users/tensorflow/gists{/gist_id}',
    'starred_url': 'https://api.github.com/users/tensorflow/starred{/owner}{/repo}',
    'subscriptions_url': 'https://api.github.com/users/tensorflow/subscriptions',
    'organizations_url': 'https://api.github.com/users/tensorflow/orgs

### Reponse and Error Handling

In [50]:
from uplink import response_handler, error_handler, Field

def raise_for_status(response):
    if 200 <= response.status_code < 300:
        # Pass through the response.
        return response
    
    print(response.text)
    raise Exception(f"{response.url} got status code: {response.status_code}")

class Github(Consumer):
    @returns.json
    @response_handler(raise_for_status)
    @post('/user/repo')
    def create_repo(self, name: Field):
        """ Create user repo """
        pass

In [51]:
gh = Github(base_url="https://api.github.com/")
gh.create_repo("helloworld")

{
  "message": "Not Found",
  "documentation_url": "https://docs.github.com/rest",
  "status": "404"
}


Exception: https://api.github.com/user/repo got status code: 404

| Feature                      | `response_handler`                                 | `error_handler`                                                        |
| ---------------------------- | -------------------------------------------------- | ---------------------------------------------------------------------- |
| When does it run?            | Only **after** a response is successfully returned | When a **network/client error** occurs and **no response is returned** |
| What errors can it handle?   | HTTP errors like 400, 404, 500                     | Transport errors: timeouts, DNS failures, broken connection            |
| What object does it receive? | `requests.Response`                                | `(exc_type, exc_val, traceback, request)`                              |


In [52]:
def raise_api_error(exc_type, exc_val, exc_tb):
    """Wraps client error with custom API error"""
    raise Exception(exc_val)

class GitHub(Consumer):
    @error_handler(raise_api_error)
    @post("user/repo")
    def create_repo(self, name: Field):
        """Create a new repository."""

In [61]:
@error_handler
def network_error(exc_type, exc_val, exc_tb):
    raise RuntimeError(f"Network failure : {exc_val}")

class Github(Consumer):
    @error_handler(raise_for_status)
    @network_error
    @post('/user/repo')
    def create_repo(self, name: Field):
        """ Create new Repo for user """
        pass

### Retry

In [69]:
from uplink import retry

class Github(Consumer):

    @retry(max_attempts=5)
    @get('/user/{user}')
    def get_user(self, user: str):
        """ Return user """
        pass

gh = Github(base_url="https://fake-github.com")
gh.get_user(user="hako")


ConnectionError: HTTPSConnectionPool(host='fake-github.com', port=443): Max retries exceeded with url: /user/hako (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x10bf12210>: Failed to resolve 'fake-github.com' ([Errno 8] nodename nor servname provided, or not known)"))

In [71]:
# https://uplink.readthedocs.io/en/stable/dev/decorators.html#retry-api

class GitHub(Consumer):
    @retry(
      # Retry on 503 response status code or any exception.
      when=retry.when.status(503) | retry.when.raises(Exception),
      # Stop after 5 attempts or when backoff exceeds 10 seconds.
      stop=retry.stop.after_attempt(5) | retry.stop.after_delay(10),
      # Use exponential backoff with added randomness.
      backoff=retry.backoff.jittered(multiplier=0.5)
    )
    @get("user/{username}")
    def get_user(self, username):
      """Get user by username."""
